src/terminal/cli/prompt.lua

```diff
@@ -0,0 +1,125 @@
+-- The user can confirm their choices by pressing <Enter>
+-- Cancel their choices by pressing <Esc>
+-- @classmod cli.Prompt
+-- @usage
+-- local prompt = Prompt {
+--     prompt = "Enter something: ",
+--     value = "Hello, 你-好 World 🚀!",
+--     max_length = 62,
+--     overflow = "wrap" -- or "scroll"
+--     position = 2,
+--     fsleep = sys.sleep,
+-- }
+-- local result, exitkey = pr:run()
+
+-- todo: use the new keymap -- check cli.select
+
+local t = require("terminal")
+local sys = require("system")
+local utils = require("terminal.utils")
+local width = require("terminal.text.width")
+local output = require("terminal.output")
+local UTF8EditLine = require("terminal.text.utf8edit").UTF8EditLine
+
+-- Key bindings
+local keys = t.input.keymap.get_keys()
+local keymap = t.input.keymap.get_keymap({
+  ["\127"] = "backspace",
+})
+
+local Prompt = utils.class()
+
+function Prompt:init(opts)
+  self.value = UTF8EditLine(opts.value or "")
+  self.prompt = opts.prompt or ""        -- the prompt to display
+  self.max_length = opts.max_length      -- the maximum length of the input
+  self.drawn_before = false              -- if the prompt has been drawn
+  self.fsleep = opts.fsleep or sys.sleep -- the sleep function to use
+end
+
+function Prompt:draw(redraw)
+  if redraw or not self.prompt_ready then
+    -- we are at start of prompt
+    self.prompt_ready = true
+    t.cursor.position.column(1)
+    output.write(tostring(self.prompt))
+  else
+    -- we are at current cursor position, move to start of prompt
+    t.cursor.position.column(width.utf8swidth(self.prompt) + 1)
+  end
+  -- write prompt & value
+  local value = tostring(self.value)
+  output.write(value)
+  -- clear remainder of input size
+  output.write(string.rep(" ", self.max_length - self.value.olen))
+  -- move to cursor position
+  t.cursor.position.column(width.utf8swidth(self.prompt) + self.value.ocursor)
+  output.flush()
+end
+
+-- Read and normalize key input
+function Prompt:readKey()
+  local key = t.input.readansi(math.huge, self.fsleep)
+  return key, keymap[key] or key
+end
+
+--- Processes key input async
+-- This function listens for key events and processes them.
+-- If an exit key is pressed, it yields the input value and the exit key.
+-- @event key Triggered when a key is pressed.
+-- @tparam string key The key that was pressed.
+-- @tparam string keytype The type of the key (e.g., "ansi", "control").
+function Prompt:handleInput()
+  while true do
+    local _, keyname = self:readKey()
+    if keyname then
+      local editing_handler = self.value[keyname]
+      if editing_handler then
+        editing_handler(self.value)
+        self:draw()
+      elseif keyname == keys.escape and self.cancellable then
+        return "cancelled"
+      elseif keyname == keys.enter then
+        return "returned"
+      elseif pcall(function() return keys[keyname] end) then -- hacky way to check non-printing characters
```

Comment:
can we add a "TODO" comment that we need to fix this with a refreence to the PR that was merged in LuaSystem?